rfd/0197-prometheus-metrics.md

```diff
@@ -0,0 +1,265 @@
+---
+authors: Hugo Hervieux (hugo.hervieux@goteleport.com)
+state: draft
+---
+
+# RFD 197 - Prometheus metrics guidelines
+
+## Required Approvers
+
+* Engineering: @codingllama && @rosstimothy && @zmb3
+
+## What
+
+This RFD covers the recommended way of adding metrics to Teleport and other services such as
+tbot, the Teleport Kube Agent Updater, the Teleport Kubernetes Operator, or plugins
+(slack, teams, pagerduty, event-handler, ...).
+
+## Why
+
+Every component currently relies on the global metrics registry and package-scoped metrics.
+This poses several challenges:
+- We pick up and serve arbitrary metrics declared by any of our dependencies. This can cause registration conflicts
+  (errors, flakiness, panics), high cardinality, confusing metric names, and data loss.
+- Several Teleport components cannot run in the same go process without causing metric conflicts. This is the case in
+  integration tests (running both tbot and Teleport), or even in user-facing binaries (tctl uses `embeddedtbot`).
+- The same Teleport component cannot be started multiple times without causing metric conflicts, or inaccurate metrics.
+  For example, integration tests starting several Teleport instances will merge their metrics together.
+
+## Conflicts
+
+Metrics conflicts can happen in several cases:
+
+1. The same metric is registered with different labels by several teleport processes. This is what happens with the
+   grpc metrics currently as teleport and tbot register global metrics with the same name but different labels. This can
+   be avoided in tests by using per-process registries.
+2. The same metric is registered several times by different components. This can be avoided by namespacing and 
+   adding metrics subsystem.
+3. The same metric is registered both in the local and global registries. In this case, the process registry takes 
+   precedence, the gathering succeeds but logs errors.
+4. The same component is being started and stopped several times and re-register its metrics. This is the case for 
+   hosted plugins. We work around by creating a dedicared registry, and registering/unregistering it as a collector. 
+   See https://github.com/prometheus/client_golang/pull/1766 for an example.
+
+## Guidelines
```

Comment:
Should we add a guideline on what namespace / subsystem to use?

Should there always be a subsystem, or an empty one OK?